dev-docs/rfcs/0002-persistent-stream-flag.md

```diff
@@ -0,0 +1,75 @@
+- Start Date: 2019-03-07
+- RFC PR: [#403](https://github.com/uber/xviz/pull/403)
+- XVIZ Issue: [#64](https://github.com/uber/xviz/issues/64)
+
+# Summary
+
+Propose the addition of a per-stream metadata property that will mark the stream as being "persistent"
+to allow send-once data to be preserved for the duration of a session.  Only the most recent datum
+of the stream is required to be preserved, conforming to snapshot and incremental semantics.
+
+# Motivation
+
+The majority of XVIZ data is time-based. 
+
+Today we only have two ways to control the valid duration of stream data, one at the Streetscape.gl
+level and one at the XVIZ level.
+
+Streetscape.gl, and any XVIZ compliant viewer, can get the XVIZ Config `TIME_WINDOW` which can be
+set by the data source to determine the window of time that the viewer should inspect to find a
+stream datum.  The first datum found should be the active datum for the current viewing time.
+This is an application wide mechanism so does not allow for individual stream control.
+
+XVIZ can control invalidating a stream datum by sending an empty stream in the update message.  The
+presence of a stream with no data is interpreted as invalidating any previous state at that time. This
+works per stream, and regardless of the actual TIME_WINDOW, but requires more tracking logic possibly in
+the XVIZ conversion process.
+
+Neither of these control mechanism allow for marking data as persistent during the whole session.
+
+# Proposal
+
+In order to mark a stream as containing persistent data that should not be purged for the duration of
+of a session I propose the following.
+
+Adding a stream metadata property:
+```
+persistent: [True|False] // False is default value
+```
+
+If this property is present and **True**, then the data should be treated special by not dropping this stream
+during any purge operation of XVIZ data.  The data should also remain visible outside the TIME_WINDOW setting.
+
+Only the most recent value of the stream is required to be preserved, allowing updates to occur and old data to
+be purged while maintaining overall persistence at the stream datum level, but not at the individual stream object level.
+
+If the update type is snapshot, the data is replaced.  If the update type is incremental, the data will be added to the
```

Comment:
I think for clarities sake it would be cool to have a few tables for different use cases.  Because right now I am not sure how you you merge in new data for each specific type of object we have.  Also does that means in `incremental` mode it's impossible to get rid of persistent data?   `snapshot` I think is supposed to be a full world snapshot (I had thought of the idea of adding "domains" or something so a snapshot could just apply to part of the stream world)

When new `state_update` arrives:

| Mode | Stream Present | Stream Not Present |
| ----- | ----- | ----- | 
| Incremental | ? | ? |
| Snapshot | ?  | ? |

Reply:
I can do this, we probably need this better specified in our regular spec too.

As of right now, increment is not actually implemented.  But my understanding was that it would simply concatenate onto the snapshot array for the data type.

For persistent mode, I expect the behavior to be latest snapshot + all incremental additions until the next snapshot.  But this is just from a semantic view point, we don't have any data to really nail down any obvious issues with this.  *incremental* fundamentally means someone somewhere is doing state management and right now we don't have examples for this.

Reply:
So the update types mean:
 - `snapshot` - these are *all the streams*, consider empty any not provided
 - `incremental` - replace only the streams provided

The idea `incremental` mode is to allow streaming partial updates to the client vs. having to send everything all the time.  I have opened PR #404 to clarify this in the docs.

I think you are consider this to actually be an `append` type mode, which is not part of XVIZ right now.

Reply:
ah, yes I did.

Reply:
Added table, assuming the current semantics for snapshot and the "append" semantics for incremental.  Pending #404 resolution